### PR TITLE
Centralized network config for virtual machines

### PR DESCRIPTION
**Pull request by tervis-unikie**
## Description of changes

Have all of the network configuration (MACs, IPs and hostnames) for virtual machines in one place instead of configuration being scattered around in several files and having excessive redundancy.

This is work in progress.
There might be some overlaps with other work, those will be sorted out soon.

## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [ ] More detailed description in the commit message(s)
- [ ] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [ ] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [ ] Test procedure described (or includes tests). Select one or more:
  - [ ] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [ ] Author has added reviewers and removed PR draft status

### CODE REVIEW

File: modules/network/definition.nix [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          type = types.int;
+          description = ''
+            IP Address ID, the last octet of internal/debug network IP address
+          '';
+        };
+      };
+    };
+  in {
+    internalNetwork = mkOption {
+      description = "Internal network address";
```
Comment (Mic92) [inline modules/network/definition.nix:36]: Please clarify who's ip address this is:

i.e.

```suggestion
      description = "Internal network address of the host";
```

File: modules/network/definition.nix [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      default = "192.168.100.0";
+    };
+
+    debugNetwork = mkOption {
+      description = "Debug network address";
+      type = types.str;
+      default = "192.168.101.0";
+    };
+
+    virtualMachines = mkOption {
```
Comment (Mic92) [inline modules/network/definition.nix:47]: Ideally all virtual machines and the host would have access to this, meaning some sort of central inventory that can be imported in different places.

File: modules/network/definition.nix (hunk 1 above)
Comment (Mic92) [reply modules/network/definition.nix:36]: Ah. Hang on. This is not an ip address but a subnet!

```suggestion
      description = "Internal network subnet (/24)";
```

File: modules/network/definition.nix [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    };
+  in {
+    internalNetwork = mkOption {
+      description = "Internal network address";
+      type = types.str;
+      default = "192.168.100.0";
+    };
+
+    debugNetwork = mkOption {
+      description = "Debug network address";
```
Comment (Mic92) [inline modules/network/definition.nix:42]:
```suggestion
      description = "Debug network subnet for virtual machines (/24)";
```

File: modules/network/definition.nix (hunk 1 above)
Comment (Mic92) [reply modules/network/definition.nix:36]: Sad to see no ipv6 here :)

File: modules/network/definition.nix [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        hostName = mkOption {
+          type = types.str;
+          description = ''
+            Hostname
+          '';
+        };
+        macAddress = mkOption {
+          type = types.str;
+          description = ''
+            MAC address
```
Comment (Mic92) [inline modules/network/definition.nix:23]:
```suggestion
            MAC address. Ideally only use "Locally Administered Address Ranges" to not collide with assigned mac addresses from hardware i.e.
            - x2-xx-xx-xx-xx-xx
            - x6-xx-xx-xx-xx-xx
            - xA-xx-xx-xx-xx-xx
            - xE-xx-xx-xx-xx-xx
```

File: modules/network/definition.nix [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#
+# Module for Network Definitions
+#
+# The point of this module is to only store information about the network
+# configuration, and the logic that uses this information should be elsewhere.
+{lib, ...}: let
+  inherit (lib) mkOption types literalExpression;
+in {
+  options.ghaf.network.definition = let
+    networkSubmodule = types.submodule {
```
Comment (Mic92) [inline modules/network/definition.nix:12]:
```suggestion
    networkSubmodule = types.submodule ({config, ...}: {
```

File: modules/network/definition.nix [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      options = {
+        hostName = mkOption {
+          type = types.str;
+          description = ''
+            Hostname
+          '';
+        };
+        macAddress = mkOption {
+          type = types.str;
```
Comment (Mic92) [inline modules/network/definition.nix:21]:
```suggestion
          type = types.str;
          default = "02:00:00:01:01:${config.ipAddressId}";
```

File: modules/network/definition.nix [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            MAC address
+          '';
+        };
+        ipAddressId = mkOption {
+          type = types.int;
+          description = ''
+            IP Address ID, the last octet of internal/debug network IP address
+          '';
+        };
+      };
```
Comment (Mic92) [inline modules/network/definition.nix:32]:
```suggestion
      });
```

File: modules/network/definition.nix (hunk 6 above)
Comment (Mic92) [reply modules/network/definition.nix:21]: Saves some typing.

**Comment by Mic92** [comment]
There is some overlap with https://github.com/tiiuae/ghaf/pull/535

**Comment by tervis-unikie** [comment]
Thanks for the review. This is a draft PR, and we are about to change the approach.

**Comment by tervis-unikie** [comment]
Plans have changed, closing this PR.

**Pull request closed without merging**